CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc

```diff
@@ -0,0 +1,770 @@
+        std::make_pair("IONS", LHCInfoPerFill::IONS),
+        std::make_pair("COSMICS", LHCInfoPerFill::COSMICS),
+        std::make_pair("GAP", LHCInfoPerFill::GAP)};
+
+    static const std::pair<const char*, LHCInfoPerFill::ParticleType> s_particleTypeMap[] = {
+        std::make_pair("PROTON", LHCInfoPerFill::PROTON),
+        std::make_pair("PB82", LHCInfoPerFill::PB82),
+        std::make_pair("AR18", LHCInfoPerFill::AR18),
+        std::make_pair("D", LHCInfoPerFill::D),
+        std::make_pair("XE54", LHCInfoPerFill::XE54)};
+
+    LHCInfoPerFill::FillType fillTypeFromString(const std::string& s_fill_type) {
+      for (auto const& i : s_fillTypeMap)
+        if (s_fill_type == i.first)
+          return i.second;
+      return LHCInfoPerFill::UNKNOWN;
+    }
+
+    LHCInfoPerFill::ParticleType particleTypeFromString(const std::string& s_particle_type) {
+      for (auto const& i : s_particleTypeMap)
+        if (s_particle_type == i.first)
+          return i.second;
+      return LHCInfoPerFill::NONE;
+    }
+  }  // namespace LHCInfoPerFillPopConImpl
+
+  namespace impl {
+
+    template <>
+    LHCInfoPerFill::FillType from_string(const std::string& attributeValue) {
+      return from_string_impl<LHCInfoPerFill::FillType, &LHCInfoPerFillPopConImpl::fillTypeFromString>(
+          attributeValue, LHCInfoPerFill::UNKNOWN);
+    }
+
+    template <>
+    LHCInfoPerFill::ParticleType from_string(const std::string& attributeValue) {
+      return from_string_impl<LHCInfoPerFill::ParticleType, &LHCInfoPerFillPopConImpl::particleTypeFromString>(
+          attributeValue, LHCInfoPerFill::NONE);
+    }
+
+  }  // namespace impl
+}  // namespace cond
+
+LHCInfoPerFillPopConSourceHandler::LHCInfoPerFillPopConSourceHandler(edm::ParameterSet const& pset)
+    : m_debug(pset.getUntrackedParameter<bool>("debug", false)),
+      m_startTime(),
+      m_endTime(),
+      m_samplingInterval((unsigned int)pset.getUntrackedParameter<unsigned int>("samplingInterval", 300)),
+      m_endFillMode(pset.getUntrackedParameter<bool>("endFill", true)),
+      m_name(pset.getUntrackedParameter<std::string>("name", "LHCInfoPerFillPopConSourceHandler")),
+      m_connectionString(pset.getUntrackedParameter<std::string>("connectionString", "")),
+      m_ecalConnectionString(pset.getUntrackedParameter<std::string>("ecalConnectionString", "")),
+      m_dipSchema(pset.getUntrackedParameter<std::string>("DIPSchema", "")),
+      m_authpath(pset.getUntrackedParameter<std::string>("authenticationPath", "")),
+      m_omsBaseUrl(pset.getUntrackedParameter<std::string>("omsBaseUrl", "")),
+      m_fillPayload(),
+      m_prevPayload(),
+      m_tmpBuffer() {
+  if (pset.exists("startTime")) {
+    m_startTime = boost::posix_time::time_from_string(pset.getUntrackedParameter<std::string>("startTime"));
+  }
+  boost::posix_time::ptime now = boost::posix_time::second_clock::local_time();
+  m_endTime = now;
+  if (pset.exists("endTime")) {
```

Comment:
exists checks on the PSet are discouraged, please do a `fillDescription` instead, see more in https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp

Reply:
I've written `fillDescritption` in `LHCInfoPerLS(Fill)PopConSourceHandler` according to the twiki page but later I realised it doesn't work. The `_cfi.py` file is not created during `scram b` nor using `edmWriteConfigs`. When running the new PopCon the parameters aren't getting checked and the default values are not passed (in practice the former optional parameters bacame obligatory).

I think it's because LHCInfoPerLSPopConSourceHandler and LHCInfoPerFillPopConSourceHandler are not plugins by themselves and the parameters are not passed to them directly. They are used to instantiate a `PopConAnalyzer` template in `LHCInfoPerLS(Fill)PopConAnalyzer.cc`:
```
typedef popcon::PopConAnalyzer<LHCInfoPerLSPopConSourceHandler> LHCInfoPerLSPopConAnalyzer;
```
and the parameters passing is implemented in https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/CondCore/PopCon/interface/PopConAnalyzer.h#L22
So the `analyzer` receives, among others, nested `ParameterSet` parameter called "Source" and passes it to the `sourceHandler` and the sourceHandler's `fillDescritption` is ignored. 

I think that for the `fillDescritpion` to work in that case it would have to be written for the `LHCInfoPerLS(Fill)PopConAnalyzer` but I see a couple of difficulties with that:
- defining the `LHCInfoPerLS(Fill)PopConAnalyzer` couldn't be a simple typedef anymore. Might it create some problems?
- writing the `fillDescription` for the `LHCInfoPerLS(Fill)PopConAnalyzer` we would have to precisely describe the parameters not only for the `sourceHandler` but for the rest that arent passed to the `sourceHandler` too. I don't know what should be the requirements for these parameters so I would need some help with that.

Am I correct about this? What should we do: write the `fillDescription` for the analyzers or just go back to not using `fillDescription` in this case? From what I've seen there are no `fillDescriptions` to other PopCons.

BTW: using `fillDescription` I still have to make an `exists` check on the PSet but I changed it into `existsAs<std::string>` as suggested in the twiki page ("optional parameters" paragraph)

Reply:
Ahh yeah this is a complication :/ like this one needs the `fillPSetDescription` instead of the `fillDescription` as explained in the "A Plugin Module Using Another Helper Plugin" part of the twiki: 
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He

Reply:
Ok, I'll try to do it this way.

Regarding the `existsAs` in itself: I only changed it to `exsitsAs` from `exists` it as it was the suggested way in the twiki page:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Optional_Parameters

I don't see how it's affected by having the fillDesc but I could remove the `existsAs` by adding a default value (eg. empty string) and checking if the argument is empty instead of checking if it exists. Actually for `endTime` I'm checking both cases because I needed the empty string to work like this to be able to have it as an optional parameters of the python scripts that are running the PopCon.

Reply:
hi @JanChyczynski do you have an update to this?

Reply:
@makortel can you please clarify the points above?

Reply:
So apparently the instantiations of `popcon::PopConAnalyzer<S>` class template are the actual framework modules, and the `PopConAnalyzer` has a member variable of type `S` directly. Therefore this case is **not** "a plugin module using another helper plugin".

From this it also follows (as already mentioned in https://github.com/cms-sw/cmssw/pull/40817#discussion_r1157825190) that the current `LHCInfoPerFillPopConSourceHandler::fillDescriptions()` and `LHCInfoPerLSPopConSourceHandler::fillDescriptions()` do not get called.

What should be done instead is to
* define the `fillDescriptions(edm::ConfigurationDescriptions& descriptions)` functions in `popcon::PopConAnalyzer<S>`
* define `fillPSetDescription(edm::ParameterSetDescription& desc)` in `LHCInfoPerFillPopConSourceHandler` and `LHCInfoPerLSPopConSourceHandler`
* make `popcon::PopConAnalyzer<S>::fillDescriptions()` call `S::fillPSetDescription()`
  * it follows that either all `S` must have `fillPSetDescription()` implemented, or one adds code in `popcon::PopConAnalyzer<S>::fillDescriptions()` to have different behavior based on `S` having `fillPSetDescription()` or not

(I'd argue this would be complicated enough to warrant its own PR)

Reply:
Thanks Matti! Sure let's do this in another PR then, I can make a reminder issue about it

Reply:
Ok, I'll revert the changes with fillDescriptions for now then. 

So, should I remove the `existsAs` and replace it with just checking for a default value? I don't see how it's dependent on the fillDescriptions and don't really se how that's better then the `existsAs` check.

Reply:
Generally the experience has been that the application behavior is easier to understand when the behavior depends only on values of configuration parameters rather than also from the existence of parameters. E.g. `edmConfigDump` lists all parameters (that may describe the module's behavior enough for not having to look the code), overriding parameter values in python code can be done without explicit types on the right hand side (allows catching typos earlier), etc.

Written that, there are some legitimate cases for `exists` (or similar). To large degree it depends on the application (e.g. HLT does not work with optional parameters at all, RECO wants to avoid and asked for the static analyzer check).